File: apep_1432/v1/code/01b_fetch_fec.py
"""01b_fetch_fec.py — Fetch FEC small-dollar contributions for protest cities.

Uses concurrent requests to speed up API queries.
"""

import os
import csv
import json
import time
import urllib.request
import urllib.parse
from concurrent.futures import ThreadPoolExecutor, as_completed

# Load .env
env_path = os.path.join(os.path.expanduser("~"), "auto-policy-evals", ".env")
env_vars = {}
with open(env_path) as f:
    for line in f:
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        if '=' in line:
            k, v = line.split('=', 1)
            env_vars[k.strip()] = v.strip().strip('"').strip("'")

FEC_KEY = env_vars.get('FEC_API_KEY', 'DEMO_KEY')
DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "data")

# Read protest cities from GDELT data
protests_file = os.path.join(DATA_DIR, "ccc_protests.csv")
city_counts = {}
with open(protests_file) as f:
    reader = csv.DictReader(f)
    for row in reader:
        cs = row['city_state']
        city_counts[cs] = city_counts.get(cs, 0) + 1

# Top 30 cities with 50+ protests
top_cities = sorted([(cs, n) for cs, n in city_counts.items() if n >= 50],
                    key=lambda x: -x[1])[:30]

print(f"Top {len(top_cities)} cities selected")


def fetch_fec_city_period(city_state, period):
    """Fetch up to 500 contributions for a city-period."""
    city = city_state.split(',')[0].strip().upper()
    state = city_state.split(',')[1].strip()

    results = []
    last_index = None
    last_date = None

    for page in range(5):  # max 5 pages × 100 = 500
        params = {
            'api_key': FEC_KEY,
            'contributor_city': city,
            'contributor_state': state,
            'two_year_transaction_period': str(period),
            'max_amount': '200',
            'min_amount': '1',
            'per_page': '100',
        }
        if last_index:
            params['last_index'] = last_index
            params['last_contribution_receipt_date'] = last_date

        url = 'https://api.open.fec.gov/v1/schedules/schedule_a/?' + urllib.parse.urlencode(params)

        try:
            resp = urllib.request.urlopen(url, timeout=30)
            data = json.loads(resp.read())
        except Exception as e:
            break

        if not data.get('results'):
            break

        for r in data['results']:
            results.append({
                'contributor_city': r.get('contributor_city', ''),
                'contributor_state': r.get('contributor_state', ''),
                'contribution_receipt_date': r.get('contribution_receipt_date', ''),
                'contribution_receipt_amount': r.get('contribution_receipt_amount', 0),
                'committee_name': r.get('committee_name', ''),
                'contributor_name': r.get('contributor_name', ''),
            })

        indexes = data.get('pagination', {}).get('last_indexes', {})
        last_index = indexes.get('last_index')
        last_date = indexes.get('last_contribution_receipt_date')
        if not last_index:
            break

        time.sleep(0.3)

    return results


def main():
    periods = [2020, 2022]
    tasks = [(cs, p) for cs, _ in top_cities for p in periods]
    total = len(tasks)
    print(f"Total queries: {total}")

    all_results = []
    completed = 0

    # Use 3 parallel workers (FEC rate limit)
    with ThreadPoolExecutor(max_workers=3) as executor:
        futures = {executor.submit(fetch_fec_city_period, cs, p): (cs, p)
                   for cs, p in tasks}

        for future in as_completed(futures):
            completed += 1
            cs, p = futures[future]
            result = future.result()
            all_results.extend(result)
            if completed % 10 == 0:
                print(f"  {completed}/{total}: {len(all_results):,} contributions", flush=True)

    print(f"\nTotal contributions: {len(all_results):,}")

    # Write CSV
    outpath = os.path.join(DATA_DIR, "fec_contributions.csv")
    with open(outpath, 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=[
            'contributor_city', 'contributor_state', 'contribution_receipt_date',
            'contribution_receipt_amount', 'committee_name', 'contributor_name'
        ])
        writer.writeheader()
        writer.writerows(all_results)

    print(f"Saved to {outpath}")


if __name__ == "__main__":
    main()

File: apep_1432/v1/code/01b_fetch_fec.R
## 01b_fetch_fec.R — Fetch FEC small-dollar contributions for top protest cities
## apep_1432: Protests and Campaign Contributions (Weather IV)
## Strategy: query top 50 protest cities × 3 two-year cycles, cap at 500 records per query

source("00_packages.R")

ccc <- fread("../data/ccc_protests.csv") %>% as_tibble()

fec_key <- Sys.getenv("FEC_API_KEY", unset = "DEMO_KEY")

## Top protest cities
city_counts <- ccc %>%
  count(city_state, sort = TRUE) %>%
  filter(n >= 50)

## Extract city name and state
city_counts <- city_counts %>%
  mutate(
    city_upper = toupper(sub(",.*", "", city_state)),
    state = sub(".*,\\s*", "", city_state)
  )

## Take top 30 cities, 2 periods (2020 and 2022 — core analysis window)
top_cities <- head(city_counts, 30)
periods <- c(2020, 2022)

total_queries <- nrow(top_cities) * length(periods)
cat(sprintf("Fetching FEC data: %d cities x %d periods = %d queries\n",
            nrow(top_cities), length(periods), total_queries), flush = TRUE)

fec_data <- list()
query_num <- 0

for (p in periods) {
  for (j in seq_len(nrow(top_cities))) {
    query_num <- query_num + 1

    ## Build URL
    city_enc <- URLencode(top_cities$city_upper[j], reserved = TRUE)
    url <- sprintf(
      "https://api.open.fec.gov/v1/schedules/schedule_a/?api_key=%s&contributor_city=%s&contributor_state=%s&two_year_transaction_period=%d&max_amount=200&min_amount=1&per_page=100",
      fec_key, city_enc, top_cities$state[j], p
    )

    resp <- tryCatch(
      {
        Sys.sleep(0.4)
        r <- GET(url)
        content(r, as = "parsed")
      },
      error = function(e) NULL
    )

    if (!is.null(resp) && length(resp$results) > 0) {
      ## Get first page results
      page1 <- map_dfr(resp$results, ~ tibble(
        contributor_city = .x$contributor_city %||% NA_character_,
        contributor_state = .x$contributor_state %||% NA_character_,
        contribution_receipt_date = .x$contribution_receipt_date %||% NA_character_,
        contribution_receipt_amount = .x$contribution_receipt_amount %||% NA_real_,
        committee_name = .x$committee_name %||% NA_character_,
        contributor_name = .x$contributor_name %||% NA_character_
      ))
      fec_data[[query_num]] <- page1

      ## Get up to 4 more pages (500 total per city-period)
      last_idx <- resp$pagination$last_indexes$last_index
      last_dt <- resp$pagination$last_indexes$last_contribution_receipt_date
      for (pg in 2:5) {
        if (is.null(last_idx)) break
        url2 <- sprintf(
          "%s&last_index=%s&last_contribution_receipt_date=%s",
          url, last_idx, last_dt
        )
        resp2 <- tryCatch(
          {
            Sys.sleep(0.4)
            r2 <- GET(url2)
            content(r2, as = "parsed")
          },
          error = function(e) NULL
        )
        if (is.null(resp2) || length(resp2$results) == 0) break
        page_n <- map_dfr(resp2$results, ~ tibble(
          contributor_city = .x$contributor_city %||% NA_character_,
          contributor_state = .x$contributor_state %||% NA_character_,
          contribution_receipt_date = .x$contribution_receipt_date %||% NA_character_,
          contribution_receipt_amount = .x$contribution_receipt_amount %||% NA_real_,
          committee_name = .x$committee_name %||% NA_character_,
          contributor_name = .x$contributor_name %||% NA_character_
        ))
        fec_data[[length(fec_data) + 1]] <- page_n
        last_idx <- resp2$pagination$last_indexes$last_index
        last_dt <- resp2$pagination$last_indexes$last_contribution_receipt_date
      }
    }

    if (query_num %% 10 == 0) {
      n_so_far <- sum(sapply(fec_data, nrow))
      cat(sprintf("  Query %d/%d (%s, %d): %d contributions so far\n",
                  query_num, total_queries, top_cities$city_upper[j], p, n_so_far),
          flush = TRUE)
    }
  }
}

fec_df <- bind_rows(fec_data)
cat(sprintf("\nFEC raw contributions: %d rows\n", nrow(fec_df)), flush = TRUE)

if (nrow(fec_df) == 0) {
  stop("FATAL: FEC API returned zero contributions. Cannot proceed.")
}

## Clean
fec_df <- fec_df %>%
  mutate(
    contribution_date = as.Date(contribution_receipt_date),
    city = str_to_title(trimws(contributor_city)),
    state = toupper(trimws(contributor_state)),
    city_state = paste0(city, ", ", state),
    amount = as.numeric(contribution_receipt_amount)
  ) %>%
  filter(!is.na(contribution_date), !is.na(amount), amount > 0, amount <= 200)

cat(sprintf("FEC cleaned: %d contributions, %d cities\n",
            nrow(fec_df), n_distinct(fec_df$city_state)), flush = TRUE)

fwrite(fec_df, "../data/fec_contributions.csv")
cat("Saved FEC contributions.\n")
